Update GitHub documentation (#447)
SuffolkLITLab/docassemble-AssemblyLine-documentation: docs/github.md

![Upload a YAML file in the Playground](./assets/da_playground_upload_yaml.png)

[Create your Package and push it to GitHub](#create-a-new-repository).

### Create a new YAML file
**Summary**

[Go to the Playground](#go-to-the-playground), tap "Add", and give the file [an appropriate name](name_formats.md#form-files-stored-in-docassemble-snake_case).

**Details**

[Go to the Playground](#go-to-the-playground) and tap "Add".

![Tap "Add"](./assets/da_playground_add.png)

Give the file [an appropriate name](name_formats.md#form-files-stored-in-docassemble-snake_case). Then save the file.

![Name and save the YAML file](./assets/da_playground_save.png)

Your save options should change below the code area to indicate the save was successful.

![The editor buttons change once the file has been saved the first time](./assets/da_playground_after_save.png)
## Workflow

### Upload a YAML file
See the [docassemble docs about Playground YAML files](https://docassemble.org/docs/playground.html#interview_files).
Here is what we recommend for a workflow for using GitHub with the Docassemble playground. Repeat these steps each time you start working on a new task or issue:

> ### Upload a package

Reviewer: I think we still need to document this step as it's the first step after using the Weaver. It could link to another part of the documentation though

* Collaborate with others to build Docassemble interviews, get help, and hand off your code to future interview builders and maintainers.
* Share your code with others who might want to build similar interviews.
* Create branches to work on specific issues, or create multiple branches to keep different issues isolated from each other while you work on them.
* Track and discuss issues with your code, like tasks, bugs, feature requests, etc. Issues can be assigned, tagged, and attached to commits and pull requests that complete or fix them.
* Use projects to gather issues from multiple repositories and organize them on a Kanban-style project board.

:::info
This page covers how to use GitHub with Docassemble. Refer to the [GitHub documentation](https://docs.github.com/en) if you have questions about its features. The [Git documentation](https://git-scm.com/doc) may also be helpful if you want to understand GitHub's underlying version control software.
:::

## Workflow

Here is what we recommend for a workflow for using GitHub with the Docassemble playground. Repeat these steps each time you start working on a new task or issue:

> 1. Create a new project in your Docassemble playground. Give it a name that describes the work to be done, then pull the `main` branch of your interview from GitHub into the project.
> 2. When you make your first commit from the new project, [commit to a new branch](#commit-your-code).
> 3. When you finish working on the branch, commit it one last time, then go to the repository in GitHub and [create a pull request](#create-a-pull-request). If you are working with someone on the interview, request a review from them.
> 4. Once you have resolved any conflicts and any reviewers have signed off, merge the pull request.
> 5. Decide on the next task or issue you want to work on and start over at #1.

Author: ```suggestion
1. Create a new project in your Docassemble playground. Give it a name that describes the work to be done, then pull the `main` branch of your interview from GitHub into the project.
2. When you make your first commit from the new project, [commit to a new branch](#commit-your-code).
3. When you finish working on the branch, commit it one last time, then go to the repository in GitHub and [create a pull request](#create-a-pull-request). If you are working with someone on the interview, request a review from them.
4. Once you have resolved any conflicts and any reviewers have signed off, merge the pull request.
5. Decide on the next task or issue you want to work on and start over at #1.

:::warning
When you pull a GitHub repository to a playground project, the files in the repository will overwrite any files with the same name in your project **without warning**. Create a new project, first, if you don't want to risk this.
:::
```

:::tip
In order to transfer a repository to an organization you must be a member of the organization and have permission to create a repository in the organization.
:::

### Create and Manage Playground Projects

To create and manage projects, in the Docassemble playground click **Playground** in the nav menu and select **Manage Projects**.

:::tip
You must use [PascalCase](naming#pascalcase) when naming projects.
:::

### Commit Your Code

1. In the Docassemble playground, click on the **Folders** menu and select **Packages**.
2. Scroll to the bottom of the page and click on the **GitHub** button.
3. Select the branch you want to commit to, enter a [commit message](#branches-pull-requests-and-commits), and click the **Commit** button. 

That's it! Your commit should show up on GitHub!

### Create a Pull Request

> Refer to the [GitHub documentation for how to create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

Author: I agree, but I also think it makes sense to see this in the sidebar:

> **How To**
> * Commit Your Code
> * Create a Pull Request
> * Resolve Conflicts
> * Review Pull Requests

And the section on reviewing pull requests is for reviewers, not for creating it in the first place. So I think I'm going to leave it for now, but I will keep it in mind for future updates.

I'm sure the next cohort of interview builders will run into different problems that I can use as prompts for future updates.

### Resolve Conflicts

> Refer to the [GitHub documentation on resolving merge conflicts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github).

Author: I don't think I know enough about how merge conflicts change the head branch before the merge, but I'd like to. Can you give me the short version, perhaps in Teams?

> Refer to the [GitHub documentation on resolving merge conflicts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github).

Reviewer: I think what Michelle means is that when you use the merge editor on GitHub, it merges `main` into your feature branch. The pull request metaphor does break down a little when you do a manual edit to resolve a conflict. That is a bit surprising compared to how other edits work in a branch in GitHub, but it's a feature of Git--"pull requests" are just a fancy GitHub-ism for "merge main", and aren't a concept that Git has natively. 

Personally I think this is too nuanced for this documentation and isn't specific enough to Docassemble to go in our documentation. Unless I'm misunderstanding what Michelle is discussing.

You can also [comparing commits, branches, etc.](https://docs.github.com/en/github/committing-changes-to-your-project/comparing-commits) in GitHub at any time.
:::

### Review Pull Requests

When someone assigns you to review a pull request (PR), you will be able to see it in [your GitHub notifications](https://github.com/notifications). Don't let these requests sit around too long, or you increase the likelihood of [merge conflicts](#resolve-conflicts).

The requester should have left notes on what needs to be tested in the description of the pull request.

1. Pull the code into the a new playground project.
2. Test the change(s) made in the pull request.
3. If it does not do what it is supposed to do, leave a comment requesting changes.
4. If it does what it is supposed to do, approve the changes and then you or the PR submitter can merge the PR and delete the branch.

:::tip
You can [make comments connected to specific lines of code](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request). This is often helpful when reviewing PRs.
:::

> Afterward you can delete the playground project you created for the review.

Author: ```suggestion
Then, delete the playground project you created for the review.
```

1. As described in that documentation, select 'Request Changes'.
1. Let the requester know about your review.

Finally, [delete the Project](https://docassemble.org/docs/playground.html#projects) you created for the review.

<!-- Clarifying: Comparing the code on GitHub   -->
<!-- Pictures from GitHub as to where to see the comparison. -->
<!-- TODO: Add pictures -->

:::note
1. You can [make comments connected to specific lines of code](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request).
1. You can make a pull request with someone's branch even if the person is not there as long as you have write permissions on the repository.
:::

### Get GitHub code into your Playground

aka. Uploading your code to the Playground, uploading someone else's code to the Playground, reviewing, pulling, getting your old code back, seeing someone else's code, working on someone else's code

:::warning

> **WARNING:** All the files in the repository will replace files of the same name in this Project. Imagine being on your computer, copying a bunch of files to a folder, and getting the message "This file already exists in this folder. Do you want to replace it?". It's the same thing, but docassemble doesn't give you that warning.

Author: Good catch. There isn't an obvious place for it here, so until I have a chance to write more about pulling from GitHub I added this warning to the Workflow section so it won't get missed.

> As a general rule the `main` branch should only be updated with pull requests from branches.

Author: ```suggestion
As a general rule, avoid editing the branch called `main`, the default branch. Only update `main` with pull requests from other branches. As another general rule, avoid creating branches from branches other than `main`—this is a recipe for merge conflicts.
```

> * Since branches usually precede pull requests in the [workflow](#workflow), branch names should anticipate the future pull request title.
> * While you are working on a branch, [commit early and often](#commit-early--often), and use the commit message to describe the change you just made to the interview code. Examples of good commit messages:
> * Pull request titles should describe the bigger change, and pull request descriptions should provide context, mention related issues, and generally explain the reason for the change.

Author: ```suggestion
* Branch names are usually short. They are a description of the purpose of the changes you will make—a description that you or your teammates can understand at a glance.
* While you are working on a branch, [commit early and often](#commit-early--often), and use the commit message to describe the change you just made to the interview code.
* Pull request titles should be a summary of the purpose of your changes. They are usually more detailed than the branch name. Pull request descriptions should provide context, [mention related issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#about-linked-issues-and-pull-requests), and generally explain the reason for the change.
```
